qa/python_models/dlpack_test/model.py

```diff
@@ -100,27 +119,50 @@ def test_dlpack_gpu_tensors(self):
             # the same
             pytorch_tensor_dlpack = from_dlpack(pb_tensor.to_dlpack())
             self.assertTrue(torch.all(pytorch_tensor_dlpack == pytorch_tensor))
-
-            # DLPack does not properly support bool type:
-            # https://github.com/google/jax/issues/4719
-            if pytorch_dtype != torch.bool:
-                self.assertTrue(
-                    pytorch_tensor.type() == pytorch_tensor_dlpack.type())
-            else:
-                self.assertFalse(
-                    pytorch_tensor.type() == pytorch_tensor_dlpack.type())
+            self.assertTrue(
+                pytorch_tensor.type() == pytorch_tensor_dlpack.type())
+            
+            # Now we make sure that updated DLPack implementation works
+            # with GPU as well
+            pb_tensor = pb_utils.Tensor.from_dlpack('test_tensor',
+                                                    pytorch_tensor)
+            pytorch_tensor_dlpack = from_dlpack(pb_tensor)
+            self.assertTrue(torch.all(pytorch_tensor_dlpack == pytorch_tensor))
+            self.assertTrue(
+                pytorch_tensor.type() == pytorch_tensor_dlpack.type())
+            
 
     def test_dlpack_gpu_numpy(self):
         # DLPack tesnors that are in GPU cannot be converted to NumPy
         pytorch_tensor = torch.rand([100], dtype=torch.float16,
                                     device='cuda') * 100
         pb_tensor = pb_utils.Tensor.from_dlpack('tensor',
                                                 to_dlpack(pytorch_tensor))
+        # Make sure that `__dlpack_device__` works as expected
+        self.assertTrue(pb_tensor.__dlpack_device__() == pytorch_tensor.__dlpack_device__())
+    
         with self.assertRaises(Exception) as e:
             pb_tensor.as_numpy()
         self.assertTrue(
             str(e.exception) ==
             'Tensor is stored in GPU and cannot be converted to NumPy.')
+    
+    def test_dlpack_cpu_numpy(self):
+        # Check compatibiity of PbTensor DLPack implementation
+        # with numpy
+        pytorch_tensor = torch.rand([100], dtype=torch.float16,
+                                    device='cpu') * 100
+        pb_tensor = pb_utils.Tensor.from_dlpack('tensor', pytorch_tensor)
+        numpy_tensor_dlpack = np.from_dlpack(pb_tensor)
+        self.assertTrue(np.all(numpy_tensor_dlpack == pytorch_tensor.numpy()))
+        # Make sure that `__dlpack_device__` works as expected
+        self.assertTrue(pb_tensor.__dlpack_device__() == pytorch_tensor.__dlpack_device__())
+    
+    def test_pdtensor_bool_internal_support(self):
+        bool_array = np.asarray([False, True])
+        bool_tensor = pb_utils.Tensor('tensor', bool_array)
+        bool_tensor_dlpack = pb_utils.Tensor.from_dlpack('tensor', bool_tensor)
```

Comment:
Would it be possible to test passing the numpy object directly in `from_dlpack`? For example,

pb_utils.Tensor.from_dlpack('tensor', bool_array)?

Reply:
Not yet. Released Numpy libraries are not supporting bool types by their implementation of DLPack. So for now, if you directly pass `bool_array` to `from_dlpack`, it will cause `TypeError: DLPack only supports signed/unsigned integers, float and complex dtypes.`  Bool support for DLPack was added 2 months ago to master numpy brunch and one of future releases should support that. I can add FIXME to change it later.

Reply:
I see. Thanks.